iroh/src/ticket.rs

```diff
@@ -0,0 +1,93 @@
+//! This module manages the different tickets Iroh has.
+
+pub mod blob;
+pub mod doc;
+
+/// Kind of ticket.
+#[derive(Debug, strum::EnumString, strum::Display, PartialEq, Eq, Clone, Copy)]
+#[strum(serialize_all = "snake_case")]
+pub enum Kind {
+    /// A blob ticket.
+    Blob,
+    /// A document ticket.
+    Doc,
+    /// A ticket for an Iroh node.
+    Node,
+}
+
+impl Kind {
+    /// Parse the ticket prefix to obtain the [`Kind`] and remainig string.
+    pub fn parse_prefix(s: &str) -> Option<Result<(Self, &str), Error>> {
+        let (prefix, rest) = s.split_once(':')?;
+        match prefix.parse() {
+            Ok(kind) => Some(Ok((kind, rest))),
+            Err(e) => Some(Err(e.into())),
+        }
+    }
+}
+
+/// An error deserializing an iroh ticket.
+#[derive(Debug, derive_more::Display, thiserror::Error)]
+pub enum Error {
+    /// Found a ticket of the wrong [`Kind`].
+    #[display("expected a {expected} ticket but found {found}")]
+    WrongKind {
+        /// Expected [`Kind`] of ticket.
+        expected: Kind,
+        /// Found [`Kind`] of ticket.
+        found: Kind,
+    },
+    /// It appears to be a ticket but the prefix is not a known one.
+    #[display("unrecogized ticket prefix")]
+    UnrecognizedKind(#[from] strum::ParseError),
+    /// This does not appear to be a ticket.
+    #[display("not a {expected} ticket")]
+    MissingKind {
+        /// Prefix that is missing.
+        expected: Kind,
+    },
+    /// This looks like a ticket, but postcard deserialization failed.
+    #[display("deserialization failed: {_0}")]
+    Postcard(#[from] postcard::Error),
+    /// This looks like a ticket, but base32 decoding failed.
+    #[display("decoding failed: {_0}")]
+    Encoding(#[from] data_encoding::DecodeError),
+}
+
+trait IrohTicket: serde::Serialize + for<'de> serde::Deserialize<'de> {
+    /// Kind of Iroh ticket.
+    const KIND: Kind;
+
+    /// Serialize to postcard bytes.
+    fn to_bytes(&self) -> Vec<u8> {
+        postcard::to_stdvec(&self).expect("postcard::to_stdvec is infallible")
+    }
+
+    /// Deserialize from postcard bytes.
+    fn from_bytes(bytes: &[u8]) -> Result<Self, postcard::Error> {
+        postcard::from_bytes(bytes)
+    }
+
+    /// Serialize to string.
+    fn serialize(&self) -> String {
+        let mut out = Self::KIND.to_string();
+        out.push(':');
```

Comment:
ok, I want to propose we use `_` instead of `:` to separate. my reason: I can't double-click the ticket to get the entire string if we use the colon. If we use the underscore it'll work

Reply:
curious I thought this would happen but weirdly it works on my terminal